File: app/src/main/java/com/pkk/andriod/attendence/activity/StudentActivity.java
package com.pkk.andriod.attendence.activity;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.wifi.p2p.WifiP2pInfo;
import android.net.wifi.p2p.WifiP2pManager;
import android.os.Looper;

import androidx.appcompat.app.AppCompatActivity;
import androidx.cardview.widget.CardView;

import android.os.Bundle;
import android.text.InputType;
import android.util.Log;
import android.widget.TextView;
import android.os.Handler;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.pkk.andriod.attendence.R;
import com.pkk.andriod.attendence.AsyncTasks.ClientAsyncSocket;
import com.pkk.andriod.attendence.fragment.PeerSelectorFragment;
import com.pkk.andriod.attendence.misc.MessageModel;


public class StudentActivity extends AppCompatActivity implements View.OnClickListener, PeerSelectorFragment.PeerFragmentToStudentActivity {

    public TextView mTextViewReplyFromServer;
    private EditText mEditTextSendMessage;
    public Button status;
    private String rollNo;
    private Boolean check = true;
    private String serverIP;
    private CardView coverCard;
    private WifiP2pManager manager;
    private WifiP2pManager.Channel channel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_student);

        coverCard = findViewById(R.id.student_activity);
        Intent intent = getIntent();
        if (intent != null)
            serverIP = intent.getStringExtra("ServerIP");

        if (manager == null && serverIP == null) {
            manager = (WifiP2pManager) getSystemService(Context.WIFI_P2P_SERVICE);
            channel = manager.initialize(this, Looper.getMainLooper(), null);
        } else if (serverIP == null) {
            manager.requestConnectionInfo(channel, new WifiP2pManager.ConnectionInfoListener() {
                @Override
                public void onConnectionInfoAvailable(WifiP2pInfo wifiP2pInfo) {
                    serverIP = wifiP2pInfo.groupOwnerAddress.getHostAddress();
                }
            });
        }
        if (serverIP == null || serverIP.isEmpty()) {
            coverCard.setVisibility(View.VISIBLE);
            PeerSelectorFragment fragment = new PeerSelectorFragment();
            fragment.setManager(manager, channel);
            getSupportFragmentManager().beginTransaction().replace(R.id.student_activity, fragment).addToBackStack(null).commit();
        } else {
            coverCard.setVisibility(View.GONE);
            Log.d("Serverip: ", serverIP);
            mEditTextSendMessage = findViewById(R.id.edt_send_message);
            mTextViewReplyFromServer = findViewById(R.id.tv_reply_from_server);
            Button buttonSend = findViewById(R.id.btn_send);
            status = findViewById(R.id.status);

            status.setOnClickListener(this);
            buttonSend.setOnClickListener(this);
        }
    }

    @Override
    public void onClick(View v) {

        switch (v.getId()) {
            case R.id.btn_send:
                MessageModel m = new MessageModel(1, mEditTextSendMessage.getText().toString());
                new ClientAsyncSocket(StudentActivity.this, serverIP, m).execute();
                break;

            case R.id.status:
                AlertDialog.Builder builder = new AlertDialog.Builder(this);
                builder.setTitle("Enter Your Roll No");
                // Set up the input
                final EditText input = new EditText(this);
                // Specify the type of input expected; this, for example, sets the input as a password, and will mask the text
                input.setInputType(InputType.TYPE_CLASS_NUMBER);
                builder.setView(input);
                // Set up the buttons
                builder.setPositiveButton("OK", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        rollNo = input.getText().toString();
                        if (!rollNo.isEmpty()) {
                            if (check) {
                                MessageModel model = new MessageModel();
                                model.setRoll_no(Integer.parseInt(rollNo));
                                model.setPresent(true);
                                new ClientAsyncSocket(StudentActivity.this, serverIP, model).execute();
                            }
                        } else
                            toast("Enter Roll No to mark your attendence");

                    }
                });
                builder.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.cancel();
                    }
                });
                builder.show();
        }
    }

    public void toast(final String message) {
        Handler handler = new Handler(Looper.getMainLooper());

        handler.post(new Runnable() {

            @Override
            public void run() {
                Toast.makeText(getApplicationContext(), message, Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    public void sendObjects(WifiP2pManager manager, WifiP2pManager.Channel channel) {
        this.manager = manager;
        this.channel = channel;
    }
}

File: app/src/main/java/com/pkk/android/attendance/activities/StudentActivity.kt
package com.pkk.android.attendance.activities

import android.app.AlertDialog
import android.content.Intent
import android.os.Bundle
import android.text.InputType
import android.util.Log
import android.view.View
import android.widget.EditText
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.setPadding
import com.google.gson.Gson
import com.pkk.android.attendance.R
import com.pkk.android.attendance.fragments.PulseLayoutFragment
import com.pkk.android.attendance.interfaces.PassDataListener
import com.pkk.android.attendance.misc.Utils.Companion.showShortToast
import com.pkk.android.attendance.models.MessageCodes
import com.pkk.android.attendance.models.MessageModel
import kotlinx.android.synthetic.main.activity_student.*
import java.net.Inet4Address
import java.net.NetworkInterface
import java.net.SocketException

class StudentActivity : AppCompatActivity(), View.OnClickListener, PassDataListener {
    private var rollNo: String? = null
    private var gson: Gson? = null
    private var message: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_student)
        initialise()
    }

    private fun initialise() {
        status!!.setOnClickListener(this)
        buttonSend.setOnClickListener(this)
        gson = Gson()

    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.buttonSend -> {
                val m = MessageModel(MessageCodes.CUSTOM, sendMessageEditText!!.text.toString())
                message = gson!!.toJson(m)
                openFragment()
            }
            R.id.status -> {
                val builder = AlertDialog.Builder(this)
                builder.setTitle("Enter Your Roll No")

                // Set up the input
                val input = EditText(this)
                input.inputType = InputType.TYPE_CLASS_NUMBER
                input.setPadding(20)
                builder.setView(input)

                // Set up the buttons
                builder.setPositiveButton("OK") { _, _ ->
                    rollNo = input.text.toString()
                    if (rollNo!!.isNotEmpty()) {
                        val model = MessageModel()
                        model.rollNo = rollNo!!.toInt()
                        model.isPresent = true
                        model.ip = localIpAddress
                        message = gson!!.toJson(model)
                        openFragment()
                    } else showShortToast(this, "Enter Roll No to mark your attendance")
                }
                builder.setNegativeButton("Cancel") { dialog, _ -> dialog.cancel() }
                builder.show()
            }
        }
    }

    private fun openFragment() {
        try {
            supportFragmentManager.beginTransaction()
                .replace(R.id.empty_layout, PulseLayoutFragment.newInstance(message))
                .addToBackStack(null).commit()
        } catch (e: Exception) {
            Log.e("error", "exception is $e")
        }
    }

    private fun updateUI(message: String?) {
        val model = gson!!.fromJson(message, MessageModel::class.java)
        when (model.messageCodes) {
            MessageCodes.NORMAL -> {
                status!!.text = getString(R.string.present_marked)
                status!!.isClickable = false
                status!!.setBackgroundResource(R.drawable.start_button_background)
            }
            MessageCodes.CUSTOM -> {
                val s = replyFromServerTextView!!.text.toString()
                if (s == "null")
                    return
                replyFromServerTextView!!.append(
                    String.format(
                        "$s\n From Server : %s",
                        model.message
                    )
                )
            }
            else -> showShortToast(
                this, model.messageCodes.message
            )
        }
    }

    companion object {
        val localIpAddress: String?
            get() {
                try {
                    val en = NetworkInterface.getNetworkInterfaces()
                    while (en.hasMoreElements()) {
                        val intf = en.nextElement()
                        val enumIpAddr = intf.inetAddresses
                        while (enumIpAddr.hasMoreElements()) {
                            val inetAddress = enumIpAddr.nextElement()
                            if (!inetAddress.isLoopbackAddress && inetAddress is Inet4Address) {
                                return inetAddress.getHostAddress()
                            }
                        }
                    }
                } catch (ex: SocketException) {
                    ex.printStackTrace()
                }
                return null
            }
    }

    override fun passData(string: String?) {
        updateUI(string)
    }

    override fun onBackPressed() {
        if (supportFragmentManager.findFragmentById(R.id.empty_layout) == null)
            startActivity(Intent(this, HomeActivity::class.java))
        super.onBackPressed()
    }

}